lib/note-revisions/index.tsx

```diff
@@ -0,0 +1,65 @@
+import React, { Component } from 'react';
+import { connect } from 'react-redux';
+import { negate } from 'lodash';
+
+import NotePreview from '../components/note-preview';
+import TagChip from '../components/tag-chip';
+import isEmailTag from '../utils/is-email-tag';
+import { tagHashOf } from '../utils/tag-hash';
+
+import type * as S from '../state';
+import type * as T from '../types';
+
+type OwnProps = {
+  noteId: T.EntityId;
+  note?: T.Note;
+};
+
+type StateProps = {
+  allTags: Map<T.TagHash, T.Tag>;
+  noteId: T.EntityId | null;
+  note: T.Note | null;
+};
+
+type Props = OwnProps & StateProps;
+
+export class NoteRevisions extends Component<Props> {
+  static displayName = 'NoteRevisions';
+
+  getTagName = (tag: T.TagName) => {
+    const { allTags } = this.props;
+    return allTags.get(tagHashOf(tag))?.name;
+  };
```

Comment:
there's no need to have this function inside a component, and it might end up being a regret in the long-run. we can pass it in via `mapStateToProps` or better yet, pass in the tag names separately. that would leave this component quite straight-forward and remove the need to jump back and forth between the component and the wrapper to figure out what to send it.

```ts
type StateProps = {
	tagNames: String[];
}

const mapStateToProps = (state, props) => ae
	…

	return {
		noteId,
		note,
		tagNames: selectors.noteTags(state, note)
	}
}
```

then we add this selector

```ts
export const noteTags = (state, note) =>
	(note?.tags ?? [])
		.filter(tagName => !isEmailTag(tagName))
		map(tagName => canonicalTagName(state, tagName));

export const canonicalTagName = (state, tagName) =>
	state.data.tags.get(tagHashOf(tagName)).name;
```

a couple of steps back though I think this is pointing to a weakness in our model where we have to do any of this. we could/should consider transforming this when we load the note into state so we don't have to recalculate it every time.

Reply:
>there's no need to have this function inside a component, and it might end up being a regret in the long-run. we can pass it in via mapStateToProps or better yet, pass in the tag names separately. that would leave this component quite straight-forward and remove the need to jump back and forth between the component and the wrapper to figure out what to send it.

I totally agree, the tag names should be calculated through selectors instead within the component. I'll update this in the `TagField` component too.

> a couple of steps back though I think this is pointing to a weakness in our model where we have to do any of this. we could/should consider transforming this when we load the note into state so we don't have to recalculate it every time.

Do you mean that we should calculate the canonical tag names on the fly when a note is loaded?

I'm a bit confused with the use of hashes for tags, if we use the hash of the tag as the id, shouldn't we use that id in the notes to reference the tags?

Reply:
> Do you mean that we should calculate the canonical tag names on the fly when a note is loaded?

yes but I didn't mean to imply we have to do it in this PR - more of a note to us all that we're straining under this mismatch between our internal model and the external model of notes and tags. granted, tags have always been weird.

> I'm a bit confused with the use of hashes for tags, if we use the hash of the tag as the id, shouldn't we use that id in the notes to reference the tags?

Tags are weird 🤷‍♂️. The easy direction to convert is from name to hash and so there's not fundamental challenge to storing the name in the note that refers back to a tag hash. If we store the tag hash in the note itself that could end up causing unexpected "garbage" for the note model outside of Simplenote, even popping up in places like exports if we're not careful. We made the decision that we want to display the canonical names for tags in the UI but there's a fundamental recognition we have to hold that it's possible that we get out of sync with that in the note, so a best-effort is made to store the most-recently-known canonical tag name in the note list.

We have made some attempts to give tags an actual name-agnostic id and there were other challenges there. For now the tag hash has resolved the worst offenses, including things like having duplicate-but-separate tags.

Reply:
>yes but I didn't mean to imply we have to do it in this PR - more of a note to us all that we're straining under this mismatch between our internal model and the external model of notes and tags. granted, tags have always been weird.

Ok, makes sense, actually I wasn't sure if this was something that could potentially be done in this PR so thanks for clarifying it 🙇 .

>Tags are weird 🤷‍♂️. The easy direction to convert is from name to hash and so there's not fundamental challenge to storing the name in the note that refers back to a tag hash. If we store the tag hash in the note itself that could end up causing unexpected "garbage" for the note model outside of Simplenote, even popping up in places like exports if we're not careful. We made the decision that we want to display the canonical names for tags in the UI but there's a fundamental recognition we have to hold that it's possible that we get out of sync with that in the note, so a best-effort is made to store the most-recently-known canonical tag name in the note list.

>We have made some attempts to give tags an actual name-agnostic id and there were other challenges there. For now the tag hash has resolved the worst offenses, including things like having duplicate-but-separate tags.

Tags can be tricky to manage for sure, thank you very much for the explanation because it helped me to better understand how they work 👍 .

A name-agnostic id is the first thing that came to my mind when I was reviewing this part, however since tag names should be unique probably it's not a trivial thing to incorporate it.

For now my solution has been to add a selector similar to what you posted in the previous comment, this way at least we won't have duplicated code and besides it simplifies the rendering of the tags ([commit](https://github.com/Automattic/simplenote-electron/pull/2817/commits/0ad3a975ac4a7305f4a1940cf7adf594148a663d)).